app/lib/service/entrypoint/_isolate.dart

```diff
@@ -139,23 +123,67 @@ class IsolateGroup {
     }
   }
 
-  /// Starts [count] new isolates, and after a [wait] duration,
-  /// closes the old ones.
+  /// Starts [count] new isolates, waits for the pending requests to get processed,
+  /// and after a maximum [wait] duration, closes the old ones.
   Future<void> renew({
     required int count,
     required Duration wait,
   }) async {
     final isolatesToClose = [..._isolates];
+    // mark the current isolates, so that they don't trigger automatic restart
     for (final i in isolatesToClose) {
-      i.shouldRestart = false;
+      i.markedForReplace = true;
     }
+    // start new isolates
     await start(count);
-    await Future.delayed(wait);
+
+    // with short breaks close the isolates that have no pending requests
+    final sw = Stopwatch()..start();
+    while (isolatesToClose.isNotEmpty && sw.elapsed < wait) {
+      if (isolatesToClose.last._pendingRequests == 0) {
+        final last = isolatesToClose.removeLast();
+        await last.close();
+      } else {
+        await Future.delayed(Duration(milliseconds: 100));
+      }
+    }
+
+    // close the remaining ones
     for (final i in isolatesToClose) {
       await i.close();
     }
   }
 
+  /// Process a request message by delegating it to one if the running isolates,
+  /// preferably one that is not under renewal.
+  void processRequestMessage(RequestMessage e) {
+    if (_isolates.isEmpty) {
+      logger.warning('No isolate to process request.');
+      e.replyPort.send(
+          ReplyMessage.error('No isolate to process request.').encodeAsJson());
+      return;
+    }
+    final last = _isolates.lastWhereOrNull((i) =>
+        i.markedForReplace == false &&
+        i._readyMessage?.requestSendPort != null);
+    if (last == null) {
+      logger.warning('No active isolate to process request.');
+      e.replyPort.send(
+          ReplyMessage.error('No isolate to process request.').encodeAsJson());
+      return;
+    }
+
+    final intermediatePort = ReceivePort();
```

Comment:
More ports that are only closed if a message is actually sent to it. If a reply never arrives then it'll never be closed.